src/patchwork/annotations.ts

```diff
@@ -0,0 +1,406 @@
+        return;
+      }
+
+      // turn anchors of discussion into highlight annotations
+      const discussionHighlightAnnotations: HighlightAnnotation<
+        unknown,
+        unknown
+      >[] = (discussion.target ?? []).flatMap((anchor) => {
+        const value = valueOfAnchor(doc, anchor);
+
+        return value !== undefined
+          ? [
+              {
+                type: "highlighted",
+                target: anchor,
+                value,
+              },
+            ]
+          : [];
+      });
+
+      // ingore discussions without highlight annotations
+      // this can happen if the values that where referenced by a discussion have since been deleted
+      if (discussionHighlightAnnotations.length === 0) {
+        return;
+      }
+
+      highlightAnnotations.push(...discussionHighlightAnnotations);
+
+      const overlappingAnnotations = [];
+
+      editAnnotations.forEach((editAnnotation) => {
+        if (
+          discussion.target.some((anchor) =>
+            doAnchorsOverlap(docType, editAnnotation.target, anchor, doc)
+          )
+        ) {
+          claimedAnnotations.add(editAnnotation);
+          overlappingAnnotations.push(editAnnotation);
+        }
+      });
+
+      discussionGroups.push({
+        annotations: discussionHighlightAnnotations.concat(
+          overlappingAnnotations
+        ),
+        discussion,
+      });
+    });
+
+    const computedAnnotationGroups: AnnotationGroup<unknown, unknown>[] =
+      groupAnnotations(
+        docType,
+        editAnnotations.filter(
+          (annotation) => !claimedAnnotations.has(annotation)
+        )
+      ).map((annotations) => ({ annotations }));
+
+    const combinedAnnotationGroups = discussionGroups.concat(
+      computedAnnotationGroups
+    );
+
+    const sortAnchorsBy = docTypes[docType].sortAnchorsBy;
+
+    return {
+      annotations: editAnnotations.concat(highlightAnnotations),
+      annotationGroups: sortAnchorsBy
+        ? sortBy(combinedAnnotationGroups, (annotationGroup) =>
+            min(
+              annotationGroup.annotations.map((annotation) =>
+                sortAnchorsBy(doc, annotation.target)
+              )
+            )
+          )
+        : combinedAnnotationGroups,
+    };
+  }, [doc, discussions, diff]);
+
+  const {
+    selectedAnchors,
+    hoveredAnchors,
+    selectedAnnotationGroupIds,
+    expandedAnnotationGroupId,
+  } = useMemo(() => {
+    const selectedAnchors = new Set<unknown>();
+    const hoveredAnchors = new Set<unknown>();
+    const selectedAnnotationGroupIds = new Set<string>();
+    let expandedAnnotationGroupId: string;
+
+    // Record selection state for anchors and annotation groups
+    switch (selectedState?.type) {
+      case "anchors": {
+        // focus selected anchors
+        selectedState.anchors.forEach((anchor) => selectedAnchors.add(anchor));
+
+        // first annotationGroup that contains all selected anchors is expanded
```

Comment:
@paulsonnentag is "expanded" actually a thing that happens in the UI? not sure i've ever seen this happen but might have missed it

Reply:
With expand I mean that you can see the comment / resolve button